src/msg_recorder/recorder.cc

```diff
@@ -0,0 +1,56 @@
+#include "cris/core/msg_recorder/recorder.h"
+
+#include "cris/core/utils/logging.h"
+
+#include "impl/utils.h"
+
+#include <filesystem>
+#include <sstream>
+
+namespace cris::core {
+
+MessageRecorder::MessageRecorder(const std::filesystem::path& record_dir_prefix, std::shared_ptr<JobRunner> runner)
+    : Base(std::move(runner))
+    , record_dir_(record_dir_prefix / RecordDirNameGenerator()) {
+    std::filesystem::create_directories(record_dir_);
+}
+
+MessageRecorder::~MessageRecorder() {
+    files_.clear();
+
+    if (std::filesystem::is_empty(GetRecordDir())) {
+        LOG(INFO) << "Record dir " << GetRecordDir() << " is empty, removing...";
+        std::filesystem::remove(GetRecordDir());
+    }
+}
+
+std::string MessageRecorder::RecordDirNameGenerator() {
+    auto              time = std::time(nullptr);
+    std::stringstream dir_name;
```

Comment:
1. ostringstream
2. This is slow if you only use it once without advance formatting. Better to simply concat the string.

Reply:
Using fmtlib instead since it is much simpler.

Reply:
I would say don't abuse it, it's still a 3rd party lib added to address json's issue.
This kind of simple case only needs + and to_string

Reply:
I don't know if there are to_string for time points, but fmtlib simplifies the formatting boilerplates by a lot.